### ISSUE TITLE
Locales broken on FHSUserEnv

### ISSUE THREAD

**Issue by jluttine** [NixOS/nixpkgs#56347]
## Issue description

Recently, my FHS user env broke. I've been using it for a few years every day without problems but now some recent change in nixos-unstable broke it.

### Steps to reproduce

Here's a minimal example:

I launch the FHS user env:

```
$ nix-shell locales-test.nix 
bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
```

Inside the FHS user env, I can try to set locales and get the same error:
```
$ export LC_ALL="en_US.UTF-8"
bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
```

My minimal FHS user env definition file `locales-test.nix` for reproduction looks like:

```
{ pkgs ? import <nixpkgs> {} }:

(
  pkgs.buildFHSUserEnv {
    name = "test";
    targetPkgs = pkgs: (
      with pkgs; [
        bash
        glibcLocales
      ]
    );
  }
).env
```

For comparison, this works fine:

```
$ nix-shell --pure -p bash glibcLocales
[nix-shell:~]$ export LC_ALL="en_US.UTF-8"
```

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the
results.

```
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.102, NixOS, 19.03.git.19eedaf (Koi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2`
 - nixpkgs: `/etc/nixpkgs`
```

**Comment by FRidh** [comment]
I've been seeing these warnings as well, but I did not notice anything being broken. Might be related to `C.UTF-8`? What if you remove `LC_ALL` and `glibcLocales` and instead export `LANG=C.UTF-8`.

**Comment by jluttine** [comment]
After updating nixos-unstable branch again, I'm not able to reproduce any of my original issues.. So I guess this is ok then.

**Comment by bachp** [comment]
I found a similiar issue with `python3` that might have the same cause. It still happens on the current   [`nixos-unstable-small` (eb2a26f5c61)](https://github.com/NixOS/nixpkgs/commit/eb2a26f5c61) channel.

It can be reproduced as follows:

1. Create a file `locale-test-python.nix` with the following content:
```
{ pkgs ? import <nixpkgs> {} }:

(
  pkgs.buildFHSUserEnv {
    name = "test";
    targetPkgs = pkgs: (
      with pkgs; [
        bash
        python3
        glibcLocales
      ]
    );
  }
).env
```

2. Run `nix-shell locale-test-python.nix`
3. Run `python3 -c 'import locale; locale.setlocale(locale.LC_ALL, "en_US.UTF-8")'`

This will generate the following error:
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/nix/store/gbwx5sc4w0i0ipnmsa0nsk88svs983vl-python3-3.7.2/lib/python3.7/locale.py", line 604, in setlocale
    return _setlocale(category, locale)
locale.Error: unsupported locale setting
```

What I noticed is that setting the locale to `C.UTF-8` also works.
`3. Run `python3 -c 'import locale; locale.setlocale(locale.LC_ALL, "C.UTF-8")'``

Doing the same in `nix-shell -p python3` works and runs without error.

**Comment by bachp** [comment]
I belive this could also cause the issue mentioned in https://github.com/NixOS/nixpkgs/pull/56565#pullrequestreview-215261954

**Comment by bachp** [comment]
What I found out so far is that:

Inside the FHSEnv `local-archive` point to `/nix/store/xvj2l5llpmcgx99cl37v5xqpazs38195-glibc-multi-2.27/lib/locale/locale-archive` (Size: 1.8M)
While outside it points to `/nix/store/67b3n3zga54zic8s72vd4sdz1nl4lr0q-glibc-locales-2.27/lib/locale/locale-archive` (Size: 122M)

**Comment by bachp** [comment]
Changing the order of `glibc_multi` and `glibcLocales` in https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/build-fhs-userenv/env.nix#L40 make the `en_US.UTF-8` variant work again.

However the the `C.UTF-8` does not longer work as it seems the `glibcLocales` derivation doesn't include the `C.UTF-8` locale.

Might be an oversight of in https://github.com/NixOS/nixpkgs/pull/54485.

@Mic92 shouldn't `glibc-locales` include `C.UTF-8` too?

**Comment by bachp** [comment]
@jluttine @Mic92 Can you please reopen, as I don't think this is fixed.

**Comment by bachp** [comment]
Will be fixed once #57974 lands
